jax/interpreters/pxla.py

```diff
@@ -280,6 +280,9 @@ def axis_env_extend(name, size):
   env = {}
   write(core.unitvar, c.Tuple())
   if const_vals:
+    for val in const_vals:
+      if isinstance(val, xla.DeviceArray):
+        val.copy_to_host_async()
```

Comment:
I think in addition to these constants we might want to prefetch the literals (see line 269 above), which appear in-line in jaxpr equations (and subjaxpr equations). The best way to do that might be to iterate through all the equations and look for literals that are DeviceArray instances, then call the copy_to_host_async on them.

The `jaxpr_replicas` function around line 258 has logic for iterating through all jaxpr (and subjaxpr) equations, but in this case the logic would need to be slightly more complicated because it'd need to check `type(v) is core.Literal` and `isinstance(v.val, xla.DeviceArray)` for each `v` in `eqn.invars`.

@dougalm let us know if that sounds incorrect to you! (Context: the point here is to, at jaxpr-to-xla-computation lowering time, fetch all constants and literals to the host (which is where the compiler runs) in a way that overlaps (instead of sums) the fetch latency costs, i.e. execute async fetches instead of one-by-one blocking fetches.)

Reply:
I added code to prefetch the literals (only in pxla.py). PTAL.

Reply:
Nice! Left a comment down there; we can close this thread.